src/api/app/models/project/update_from_xml_command.rb

```diff
@@ -282,22 +282,28 @@ def update_hostsystem(current_repo, xml_hash)
       current_repo.save! if current_repo.changed?
     end
 
-    def update_repository_architectures(current_repo, xml_hash)
-      # destroy architecture references
-      Rails.logger.debug "delete all repository architectures of repository '#{project.id}'"
-      RepositoryArchitecture.where('repository_id = ?', current_repo.id).delete_all
+    def check_for_duplicated_archs!(xml_archs)
+      taken_archs = {}
 
-      position = 1
-      xml_hash.elements('arch') do |arch|
-        unless Architecture.archcache.key?(arch)
-          raise SaveError, "unknown architecture: '#{arch}'"
+      xml_archs.each do |archname|
+        if taken_archs.key?(archname)
+          raise SaveError, "double use of architecture: '#{archname}'"
         end
-        if current_repo.repository_architectures.where(architecture: Architecture.archcache[arch]).exists?
-          raise SaveError, "double use of architecture: '#{arch}'"
-        end
-        current_repo.repository_architectures.create(architecture: Architecture.archcache[arch], position: position)
-        position += 1
+        taken_archs[archname] = 1
+      end
+    end
+
+    def update_repository_architectures(current_repo, xml_hash)
+      xml_archs = xml_hash.elements('arch')
+      check_for_duplicated_archs!(xml_archs)
+
+      archs = []
+      xml_archs.each_with_index do |archname, position|
+        arch = Architecture.from_cache!(archname)
+        current_repo.repository_architectures.find_or_create_by(architecture: arch).insert_at(position + 1)
```

Comment:
The ``+ 1`` is necessary because the loop starts with zero? You could do

```ruby 
xml_archs.each.with_index(1) do |archname, position|
```

Reply:
ah, ok. The +1 is just to avoid touching one spec that hardcodes positions

Reply:
Please don't. Future us will spend cycles on trying to figure out why +1 is necessary...